plugin/execute_command.py

```diff
@@ -12,14 +13,41 @@ def __init__(self, view: sublime.View) -> None:
     def run(self,
             edit: sublime.Edit,
             command_name: Optional[str] = None,
-            command_args: Optional[Any] = None) -> None:
+            command_args: Optional[List[Any]] = None) -> None:
         client = self.client_with_capability('executeCommandProvider')
         if client and command_name:
             window = self.view.window()
             if window:
                 window.status_message("Running command {}".format(command_name))
+            if command_args:
+                command_args = self._expand_variables(command_args)
             self._send_command(client, command_name, command_args)
 
+    def _expand_variables(self, command_args: List[Any]) -> List[Any]:
+        region = self.view.sel()[-1]
+        for i, arg in enumerate(command_args):
+            if arg in ["$file_uri", "${file_uri}"]:
+                command_args[i] = uri_from_view(self.view)
+            elif arg in ["$selection", "${selection}"]:
+                command_args[i] = self.view.substr(region)
+            elif arg in ["$offset", "${offset}"]:
+                command_args[i] = region.b
+            elif arg in ["$selection_begin", "${selection_begin}"]:
+                command_args[i] = region.begin()
+            elif arg in ["$selection_end", "${selection_end}"]:
+                command_args[i] = region.end()
+            elif arg in ["$position", "${position}"]:
+                position = self.view.rowcol(region.b)
+                command_args[i] = {"line": position[0], "character": position[1]}
+            elif arg in ["$range", "${range}"]:
+                start = self.view.rowcol(region.begin())
+                end = self.view.rowcol(region.end())
+                command_args[i] = {
+                    "start": {"line": start[0], "character": start[1]},
+                    "end": {"line": end[0], "character": end[1]}
```

Comment:
Please use `region_to_range(region).to_lsp()` from plugin/core/views.py, this function does UTF-16 correctly (for ST4).